spot-client/src/common/auto-update/AutoUpdateController.js

```diff
@@ -0,0 +1,113 @@
+    setIsNightlyReloadTime,
+    updateWebSource
+} from './actions';
+import { isOkToUpdate, isTimeForNightlyReload, isWebUpdateAvailable } from './selectors';
+import TimeRangePoller from './TimeRangePoller';
+
+/**
+ * How often the controller checks if it's the nightly reload time.
+ *
+ * @type {number}
+ */
+const NIGHTLY_RELOAD_POLL_INTERVAL = 30000;
+
+/**
+ * How often the check for web source update happens.
+ *
+ * @type {number}
+ */
+const WEB_UPDATE_POLL_INTERVAL = 30000;
+
+const lastLoadTime = date.getCurrentDate();
+
+const allowedRoutes = [];
+
+// Spot TV
+allowedRoutes.push(ROUTES.SETUP);
+allowedRoutes.push(ROUTES.UNSUPPORTED_BROWSER);
+allowedRoutes.push(ROUTES.HOME);
+allowedRoutes.push(ROUTES.OLD_HOME);
+
+// Spot Remote
+allowedRoutes.push(ROUTES.CODE);
+allowedRoutes.push(ROUTES.HELP);
+
+StateListenerRegistry.register((state, prevSelection) => {
+    const enableAutoUpdate = isBackendEnabled(state) && Boolean(getPermanentPairingCode(state));
+    const currentRoute = getCurrentRoute(state);
+    const currentView = getCurrentView(state);
+
+    let isRouteAllowed = allowedRoutes.indexOf(currentRoute) !== -1;
+
+    // Special case for remote-control view
+    if (currentRoute === ROUTES.REMOTE_CONTROL && currentView !== 'meeting') {
+        isRouteAllowed = true;
+    }
+
+    const newSelection = {
+        isNightlyReloadTime: isTimeForNightlyReload(state),
+        isUpdateAllowed: enableAutoUpdate && isRouteAllowed,
+        isWebUpdateAvailable: isWebUpdateAvailable(state)
+    };
+
+    return isEqual(newSelection, prevSelection) ? prevSelection : newSelection;
+// eslint-disable-next-line no-shadow
+}, ({ isInNightlyUpdatesRange, isUpdateAllowed, isWebUpdateAvailable }, { dispatch }) => {
+    dispatch(setOkToUpdate(isUpdateAllowed));
+
+    if (isUpdateAllowed && (isWebUpdateAvailable || isInNightlyUpdatesRange)) {
+        dispatch(updateWebSource());
+    }
+});
+
+MiddlewareRegistry.register(({ dispatch, getState }) => next => action => {
+    switch (action.type) {
+    case BOOTSTRAP_COMPLETE: {
+        const state = getState();
+        const nightlyReloadTimePoller = new TimeRangePoller({
+            endHour: getUpdateEndHour(state),
+            startHour: getUpdateStartHour(state),
+            frequency: NIGHTLY_RELOAD_POLL_INTERVAL
+        });
+
+        nightlyReloadTimePoller.addListener(
+            TimeRangePoller.TIME_WITHIN_RANGE_UPDATE,
+            isTimeWithinRange => {
+                // It's time for a nightly reload if it's been more than 1 day since the last load time
+                dispatch(setIsNightlyReloadTime(isTimeWithinRange && !date.isDateForToday(lastLoadTime)));
+            }
+        );
+
+        nightlyReloadTimePoller.start();
+
+        // Checks for Web updates
+        setInterval(() => {
+            if (!isWebUpdateAvailable(getState()) && isOkToUpdate(getState())) {
+                dispatch(checkForWebUpdateAvailable());
+            }
+        },
```

Comment:
It feels to me conceptually all of this logic in the middleware could be moved into a separate class and the middleware calls into the class to handle the logic. So in other words, much like what TimeRangePoller does but with even more stuff abstracted into TimeRangePoller. I'd rather keep the integration layer between the app and "internal libs" be as thin as possible. What do you think? My main concern with jitsi middleware has always been how so much logic gets put into it, eventually leading to being unable to know what gets called when, so centralizing logic might help mitigate that (as would not have auto-registration of middleware).